Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx

```diff
@@ -92,97 +112,270 @@ function sortDialog(dialogs: DialogInfo[]) {
+type BotInProject = {
   dialogs: DialogInfo[];
-  dialogId: string;
-  selected: string;
-  onSelect: (id: string, selected?: string) => void;
+  projectId: string;
+  name: string;
+  isRemote: boolean;
+};
+
+type IProjectTreeProps = {
+  onSelect?: (link: TreeLink) => void;
+  onAllSelected?: () => void;
+  showTriggers?: boolean;
+  showDialogs?: boolean;
+  navLinks?: TreeLink[];
   onDeleteTrigger: (id: string, index: number) => void;
   onDeleteDialog: (id: string) => void;
-}
+};
 
-export const ProjectTree: React.FC<IProjectTreeProps> = (props) => {
-  const { onboardingAddCoachMarkRef } = useRecoilValue(dispatcherState);
+export const ProjectTree: React.FC<IProjectTreeProps> = ({
+  onAllSelected = undefined,
+  showTriggers = true,
+  showDialogs = true,
+  onDeleteDialog,
+  onDeleteTrigger,
+  onSelect,
+}) => {
+  const { onboardingAddCoachMarkRef, selectTo, navTo } = useRecoilValue(dispatcherState);
 
-  const groupRef: React.RefObject<IGroupedList> = useRef(null);
-  const { dialogs, dialogId, selected, onSelect, onDeleteTrigger, onDeleteDialog } = props;
   const [filter, setFilter] = useState('');
+  const [selectedLink, setSelectedLink] = useState<TreeLink | undefined>();
   const delayedSetFilter = debounce((newValue) => setFilter(newValue), 1000);
   const addMainDialogRef = useCallback((mainDialog) => onboardingAddCoachMarkRef({ mainDialog }), []);
+  const projectCollection = useRecoilValue<BotInProject[]>(botProjectSpaceSelector).map((bot) => ({
+    ...bot,
+    hasWarnings: false,
+  }));
+  const currentProjectId = useRecoilValue(currentProjectIdState);
 
-  const sortedDialogs = useMemo(() => {
-    return sortDialog(dialogs);
-  }, [dialogs]);
+  const notificationMap: { [projectId: string]: { [dialogId: string]: Diagnostic[] } } = {};
+  for (const bot of projectCollection) {
+    notificationMap[bot.projectId] = {};
+    // this seems to be the only way to get these notifications
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    const dialogMessages = useRecoilValue(validateDialogSelectorFamily(bot.projectId));
+    for (const message of dialogMessages) {
+      const dialogId = message.id;
+      notificationMap[bot.projectId][dialogId] = message.diagnostics;
+    }
+  }
+
+  const dialogHasWarnings = (dialog: DialogInfo) => {
+    notificationMap[currentProjectId][dialog.id].some((diag) => diag.severity === DiagnosticSeverity.Warning);
+  };
+
+  const botHasWarnings = (bot: BotInProject) => {
+    return bot.dialogs.some(dialogHasWarnings);
+  };
+
+  const dialogHasErrors = (dialog: DialogInfo) => {
+    notificationMap[currentProjectId][dialog.id].some((diag) => diag.severity === DiagnosticSeverity.Error);
+  };
 
-  const onRenderHeader = (props: IGroupHeaderProps) => {
-    const toggleCollapse = (): void => {
-      groupRef.current?.toggleCollapseAll(true);
-      props.onToggleCollapse?.(props.group!);
-      onSelect(props.group!.key);
+  const botHasErrors = (bot: BotInProject) => {
```

Comment:
nit: You could move it inside the selector. It would be cached once and recalculted only when dialogs change rather than being recalculated multiple times here

Reply:
If you want to keep it here, please use useMemo

Reply:
I'm not sure about that. useMemo uses reference equality to check if it needs to update, and this is a live view of whether a bot or dialog has an error. If I type something into a trigger's actions that causes an error (like "=[", which doesn't parse), this produces error icons on the trigger and the dialog it belongs to. The work it's doing is only linear in the number of triggers as well, which I don't think is costly enough to warrant useMemo.